src/util/address.rs

```diff
@@ -465,19 +455,44 @@ mod tests {
 /// operations allowed on ObjectReference are very limited. No address arithmetics
 /// are allowed for ObjectReference. The idea is from the paper
 /// High-level Low-level Programming (VEE09) and JikesRVM.
+///
+/// A runtime may define its "object references" differently. It may define an object reference as
+/// the address of an object, a handle that points to an indirection table entry where a pointer to
+/// the object is held, or anything else. Regardless, MMTk expects each object reference to have a
+/// pointer to the object (an address) in each object reference, and that address should be used
+/// for this `ObjectReference` type.
+///
+/// We currently do not allow an opaque `ObjectReference` type for which a binding can define
+/// their layout. We now only allow a binding to define their semantics through a set of
+/// methods in [`crate::vm::ObjectModel`]. Major refactoring is needed in MMTk to allow
+/// the opaque `ObjectReference` type, and we haven't seen a use case for now.
 #[repr(transparent)]
 #[derive(Copy, Clone, Eq, Hash, PartialOrd, PartialEq)]
 pub struct ObjectReference(usize);
 
 impl ObjectReference {
     pub const NULL: ObjectReference = ObjectReference(0);
 
-    /// converts the ObjectReference to an Address
+    /// Cast the object reference to its raw address. This method is mostly for the convinience of a binding.
+    ///
+    /// MMTk should not make any assumption on the actual location of the address with the object reference.
+    /// MMTk should not assume the address returned by this method is in our allocation. For the purposes of
+    /// setting object metadata, MMTk should use [`crate::vm::ObjectModel::ref_to_address()`] or [`crate::vm::ObjectModel::ref_to_header()`].
     #[inline(always)]
-    pub fn to_address(self) -> Address {
```

Comment:
Adding this would simplify the code in other files that repeatedly write `VM::VMObjectModel::ref_to_address(object)`:

```rust
pub fn to_address<VM: VMBinding>(self) -> Address {
    VM::VMObjectModel::ref_to_address(object)
}
```

Reply:
Also we can add the address->ref conversion method as well `fn from_address<VM: VMBinding>(addr: Address) -> ObjectReference`

Reply:
Good idea. I made changes as suggested.